ocaml/xapi/xapi_mgmt_iface.ml

```diff
@@ -157,20 +157,33 @@ let enable_himn ~__context ~addr =
 let rebind ~__context =
   run ~__context ~mgmt_enabled:!listening_all
 
-let management_ip_mutex = Mutex.create ()
-let management_ip_cond = Condition.create ()
+let ip_mutex = Mutex.create ()
+let ip_cond = Condition.create ()
 
 let wait_for_management_ip ~__context =
   let ip = ref (match Helpers.get_management_ip_addr ~__context with Some x -> x | None -> "") in
   let is_connected = ref (Helpers.get_management_iface_is_connected ~__context) in
-  Mutex.execute management_ip_mutex
-    (fun () -> begin while !ip = "" && !is_connected = false do
-          Condition.wait management_ip_cond management_ip_mutex;
+  Mutex.execute ip_mutex
+    (fun () -> while !ip = "" && !is_connected = false do
+          Condition.wait ip_cond ip_mutex;
           ip := (match Helpers.get_management_ip_addr ~__context with Some x -> x | None -> "");
           is_connected := (Helpers.get_management_iface_is_connected ~__context)
-        done; end);
+        done);
   !ip
 
+(* CA-280237: Called in startup sequence after creating cluster_hosts *)
+let wait_for_clustering_ip ~__context ~(self : API.ref_Cluster_host) =
+  let pIF = Db.Cluster_host.get_PIF ~__context ~self in
+  let iP = ref (Db.PIF.get_IP ~__context ~self:pIF) in
+  let attached = ref (Db.PIF.get_currently_attached ~__context ~self:pIF) in
+  Mutex.execute ip_mutex (* Don't return until PIF is plugged AND has a valid IP *)
+    (fun () -> while !iP="" || not !attached do
+        Condition.wait ip_cond ip_mutex;
+        iP := Db.PIF.get_IP ~__context ~self:pIF;
+        attached := Db.PIF.get_currently_attached ~__context ~self:pIF
```

Comment:
This does do the same as `get_management_iface_is_connected`, which is checking the PIF carrier (link state). You can use `PIF_metrics.carrier` for that.